File: forge/src/main/java/me/pandamods/pandalib/forge/platform/NetworkHelperImpl.java
/*
 * Copyright (C) 2025 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU Lesser General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package me.pandamods.pandalib.forge.platform;

import me.pandamods.pandalib.networking.NetworkContext;
import me.pandamods.pandalib.networking.NetworkReceiver;
import me.pandamods.pandalib.platform.services.NetworkHelper;
import me.pandamods.pandalib.utils.Env;
import me.pandamods.pandalib.utils.EnvRunner;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.server.level.ServerPlayer;
import net.minecraftforge.network.*;
import net.minecraftforge.network.payload.PayloadConnection;

import java.util.HashMap;
import java.util.Map;

public class NetworkHelperImpl implements NetworkHelper {
	public static final Map<CustomPacketPayload.Type<?>, Channel<CustomPacketPayload>> CHANNELS = new HashMap<>();
	private int id = 0;

	@Override
	public <T extends CustomPacketPayload> void registerClientReceiver(CustomPacketPayload.Type<T> type,
	                                                                   StreamCodec<? super RegistryFriendlyByteBuf, T> codec,
	                                                                   NetworkReceiver<T> receiver) {
		EnvRunner.runIf(Env.CLIENT, () -> () -> registerChannel(type, codec, receiver));
	}

	@Override
	public <T extends CustomPacketPayload> void registerServerReceiver(CustomPacketPayload.Type<T> type,
	                                                                   StreamCodec<? super RegistryFriendlyByteBuf, T> codec,
	                                                                   NetworkReceiver<T> receiver) {
		registerChannel(type, codec, receiver);
	}

	@Override
	public <T extends CustomPacketPayload> void registerBiDirectionalReceiver(CustomPacketPayload.Type<T> type,
	                                                                          StreamCodec<? super RegistryFriendlyByteBuf, T> codec,
	                                                                          NetworkReceiver<T> clientReceiver,
	                                                                          NetworkReceiver<T> serverReceiver) {
		registerChannel(type, codec, (ctx, payload) -> {
			switch (ctx.getDirection()) {
				case CLIENT -> clientReceiver.receive(ctx, payload);
				case SERVER -> serverReceiver.receive(ctx, payload);
				default -> throw new IllegalStateException("Unexpected value: " + ctx.getDirection());
			}
		});
	}

	@SuppressWarnings("unchecked")
	private <T extends CustomPacketPayload> void registerChannel(CustomPacketPayload.Type<T> type,
	                                                             StreamCodec<? super RegistryFriendlyByteBuf, T> codec,
	                                                             NetworkReceiver<T> receiver) {
		if (CHANNELS.containsKey(type))
			throw new IllegalArgumentException(String.format("Networking channel '%s' has already been registered", type.id()));

		PayloadConnection<CustomPacketPayload> channelBuilder = ChannelBuilder.named(type.id())
				.optional()
				.payloadChannel();

		Channel<CustomPacketPayload> channel = channelBuilder.play()
				.bidirectional()
				.addMain(type, (StreamCodec<RegistryFriendlyByteBuf, T>) codec, (msg, ctx) -> receiver.receive(new NetworkContext(
						ctx.getSender(),
						ctx.isClientSide() ? Env.CLIENT : Env.SERVER
				), (T) msg))
				.build();
		CHANNELS.put(type, channel);
	}

	@Override
	public <T extends CustomPacketPayload> void sendToServer(T payload) {
		CHANNELS.get(payload.type()).send(payload, PacketDistributor.SERVER.noArg());
	}

	@Override
	public <T extends CustomPacketPayload> void sendToPlayer(ServerPlayer player, T payload) {
		CHANNELS.get(payload.type()).send(payload, PacketDistributor.PLAYER.with(player));
	}

	@Override
	public <T extends CustomPacketPayload> void sendToAllPlayers(T payload) {
		CHANNELS.get(payload.type()).send(payload, PacketDistributor.ALL.noArg());
	}
}

File: neoforge/src/main/kotlin/platform/NetworkHelperImpl.kt
package dev.pandasystems.pandalib.neoforge.platform

import dev.architectury.utils.Env
import dev.pandasystems.pandalib.impl.networking.NetworkContext
import dev.pandasystems.pandalib.impl.networking.NetworkReceiver
import dev.pandasystems.pandalib.api.platform.NetworkHelper
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.server.level.ServerPlayer
import net.neoforged.neoforge.network.PacketDistributor
import net.neoforged.neoforge.network.event.RegisterPayloadHandlersEvent
import net.neoforged.neoforge.network.handling.DirectionalPayloadHandler
import net.neoforged.neoforge.network.handling.IPayloadContext
import net.neoforged.neoforge.network.registration.PayloadRegistrar

class NetworkHelperImpl : PayloadRegistrar("1"), NetworkHelper {
	override fun <T : CustomPacketPayload> registerClientReceiver(
		type: CustomPacketPayload.Type<T>,
		codec: StreamCodec<in RegistryFriendlyByteBuf, T>,
		receiver: NetworkReceiver<T>
	) {
		playToClient<T>(type, codec) { arg: T, ctx: IPayloadContext -> receiver.receive(NetworkContext(ctx.player(), Env.CLIENT), arg) }
	}

	override fun <T : CustomPacketPayload> registerServerReceiver(
		type: CustomPacketPayload.Type<T>,
		codec: StreamCodec<in RegistryFriendlyByteBuf, T>,
		receiver: NetworkReceiver<T>
	) {
		playToServer<T>(type, codec) { arg: T, ctx: IPayloadContext -> receiver.receive(NetworkContext(ctx.player(), Env.SERVER), arg) }
	}

	override fun <T : CustomPacketPayload> registerBiDirectionalReceiver(
		type: CustomPacketPayload.Type<T>,
		codec: StreamCodec<in RegistryFriendlyByteBuf, T>,
		clientReceiver: NetworkReceiver<T>,
		serverReceiver: NetworkReceiver<T>
	) {
		playBidirectional<T>(
			type, codec,
			DirectionalPayloadHandler<T>(
				{ arg: T, ctx: IPayloadContext -> clientReceiver.receive(NetworkContext(ctx.player(), Env.CLIENT), arg) },
				{ arg: T, ctx: IPayloadContext -> serverReceiver.receive(NetworkContext(ctx.player(), Env.SERVER), arg) }
			)
		)
	}

	override fun <T : CustomPacketPayload> sendToServer(payload: T) {
		PacketDistributor.sendToServer(payload)
	}

	override fun <T : CustomPacketPayload> sendToPlayer(player: ServerPlayer, payload: T) {
		PacketDistributor.sendToPlayer(player, payload)
	}

	override fun <T : CustomPacketPayload> sendToAllPlayers(payload: T) {
		PacketDistributor.sendToAllPlayers(payload)
	}
}
